File: RoyalVNCObjCDemo/Source/Configuration/ConfigurationViewController.m
#import "ConfigurationViewController.h"
#import "EncodingsConfigurationViewController.h"

@interface ConfigurationViewController ()

@property (strong) VNCConnectionSettings* initialSettings;

@property (strong) EncodingsConfigurationViewController* encodingsConfigurationViewController;
@property BOOL didLoad;

@property (weak) IBOutlet NSTextField* textFieldHostname;
@property (weak) IBOutlet NSTextField* textFieldPort;

@property (weak) IBOutlet NSButton* checkBoxShared;
@property (weak) IBOutlet NSButton* checkBoxClipboardRedirection;
@property (weak) IBOutlet NSButton* checkBoxScaling;
@property (weak) IBOutlet NSButton* checkBoxUseDisplayLink;
@property (weak) IBOutlet NSButton* checkBoxDebugLogging;

@property (weak) IBOutlet NSPopUpButton* popupButtonInputMode;
@property (weak) IBOutlet NSPopUpButton* popupButtonColorDepth;
@property (weak) IBOutlet NSView* placeholderViewEncodings;

@end

@implementation ConfigurationViewController

- (instancetype)initWithSettings:(VNCConnectionSettings *)settings {
	self = [super initWithNibName:@"ConfigurationView"
						   bundle:NSBundle.mainBundle];
	
	if (self) {
		self.initialSettings = settings;
		
		NSArray<NSNumber*>* supportedFrameEncodings = VNCFrameEncodingTypeUtils.defaultFrameEncodings;
		
		self.encodingsConfigurationViewController = [[EncodingsConfigurationViewController alloc] initWithSupportedFrameEncodings:supportedFrameEncodings];
	}
	
	return self;
}

- (void)viewDidLoad {
	[super viewDidLoad];
	
	if (self.didLoad) {
		return;
	}
	
	self.didLoad = YES;
	
	NSView* encodingsConfigurationView = self.encodingsConfigurationViewController.view;
	encodingsConfigurationView.frame = self.placeholderViewEncodings.bounds;
	
	[self.placeholderViewEncodings addSubview:encodingsConfigurationView];
	
	self.settings = self.initialSettings;
}

- (BOOL)becomeFirstResponder {
	return [self.textFieldHostname becomeFirstResponder];
}

- (VNCConnectionSettings *)settings {
	return [[VNCConnectionSettings alloc] initWithIsDebugLoggingEnabled:self.isDebugLoggingEnabled
															   hostname:self.hostname
																   port:self.port
															   isShared:self.isShared
													   isScalingEnabled:self.isScalingEnabled
														 useDisplayLink:self.useDisplayLink
															  inputMode:self.inputMode
										  isClipboardRedirectionEnabled:self.isClipboardRedirectionEnabled
															 colorDepth:self.colorDepth
														 frameEncodings:self.frameEncodings];
}

- (void)setSettings:(VNCConnectionSettings *)settings {
	self.isDebugLoggingEnabled = settings.isDebugLoggingEnabled;
	self.hostname = settings.hostname;
	self.port = settings.port;
	self.isShared = settings.isShared;
	self.isScalingEnabled = settings.isScalingEnabled;
	self.useDisplayLink = settings.useDisplayLink;
	self.inputMode = settings.inputMode;
	self.isClipboardRedirectionEnabled = settings.isClipboardRedirectionEnabled;
	self.colorDepth = settings.colorDepth;
	self.frameEncodings = settings.frameEncodings;
}

- (BOOL)isDebugLoggingEnabled {
	return self.checkBoxDebugLogging.state == NSControlStateValueOn;
}
- (void)setIsDebugLoggingEnabled:(BOOL)isDebugLoggingEnabled {
	self.checkBoxDebugLogging.state = isDebugLoggingEnabled ? NSControlStateValueOn : NSControlStateValueOff;
}

- (NSString*)hostname {
	return self.textFieldHostname.stringValue;
}
- (void)setHostname:(NSString*)hostname {
	self.textFieldHostname.stringValue = hostname;
}

- (uint16_t)port {
	return (uint16_t)self.textFieldPort.intValue;
}
- (void)setPort:(uint16_t)port {
	self.textFieldPort.intValue = port;
}

- (BOOL)isShared {
	return self.checkBoxShared.state == NSControlStateValueOn;
}
- (void)setIsShared:(BOOL)isShared {
	self.checkBoxShared.state = isShared ? NSControlStateValueOn : NSControlStateValueOff;
}

- (BOOL)isScalingEnabled {
	return self.checkBoxScaling.state == NSControlStateValueOn;
}
- (void)setIsScalingEnabled:(BOOL)isScalingEnabled {
	self.checkBoxScaling.state = isScalingEnabled ? NSControlStateValueOn : NSControlStateValueOff;
}

- (BOOL)useDisplayLink {
	return self.checkBoxUseDisplayLink.state == NSControlStateValueOn;
}
- (void)setUseDisplayLink:(BOOL)useDisplayLink {
	self.checkBoxUseDisplayLink.state = useDisplayLink ? NSControlStateValueOn : NSControlStateValueOff;
}

- (VNCInputMode)inputMode {
	return (VNCInputMode)self.popupButtonInputMode.indexOfSelectedItem;
}
- (void)setInputMode:(VNCInputMode)inputMode {
	[self.popupButtonInputMode selectItemAtIndex:inputMode];
}

- (BOOL)isClipboardRedirectionEnabled {
	return self.checkBoxClipboardRedirection.state == NSControlStateValueOn;
}
- (void)setIsClipboardRedirectionEnabled:(BOOL)isClipboardRedirectionEnabled {
	self.checkBoxClipboardRedirection.state = isClipboardRedirectionEnabled ? NSControlStateValueOn : NSControlStateValueOff;
}

- (VNCColorDepth)colorDepth {
	switch (self.popupButtonColorDepth.indexOfSelectedItem) {
		case 0:
			return VNCColorDepthDepth8Bit;
		case 1:
			return VNCColorDepthDepth16Bit;
		case 2:
			return VNCColorDepthDepth24Bit;
		default:
			return VNCColorDepthDepth24Bit;
	}
}
- (void)setColorDepth:(VNCColorDepth)colorDepth {
	switch (colorDepth) {
		case VNCColorDepthDepth8Bit:
			[self.popupButtonColorDepth selectItemAtIndex:0];
			break;
		case VNCColorDepthDepth16Bit:
			[self.popupButtonColorDepth selectItemAtIndex:1];
			break;
		case VNCColorDepthDepth24Bit:
			[self.popupButtonColorDepth selectItemAtIndex:2];
			break;
	}
}

- (NSArray<NSNumber*>*)frameEncodings {
	return self.encodingsConfigurationViewController.frameEncodings;
}
- (void)setFrameEncodings:(NSArray<NSNumber*>*)frameEncodings {
	self.encodingsConfigurationViewController.frameEncodings = frameEncodings;
}

@end

File: RoyalVNCDemo/Source/Configuration/ConfigurationViewController.swift
import Foundation
import AppKit

import RoyalVNCKit

class ConfigurationViewController: NSViewController {
	@IBOutlet private weak var textFieldHostname: NSTextField!
	@IBOutlet private weak var textFieldPort: NSTextField!
	
	@IBOutlet private weak var checkBoxShared: NSButton!
	@IBOutlet private weak var checkBoxClipboardRedirection: NSButton!
	@IBOutlet private weak var checkBoxScaling: NSButton!
	@IBOutlet private weak var checkBoxUseDisplayLink: NSButton!
	@IBOutlet private weak var checkBoxDebugLogging: NSButton!
	
	@IBOutlet private weak var popupButtonInputMode: NSPopUpButton!
	@IBOutlet private weak var popupButtonColorDepth: NSPopUpButton!
	@IBOutlet private weak var placeholderViewEncodings: NSView!
	
	private var didLoad = false
	private let initialSettings: VNCConnection.Settings
	private let encodingsConfigurationViewController = EncodingsConfigurationViewController(supportedFrameEncodings: .default)
	
	override var acceptsFirstResponder: Bool { true }
	
	init(settings: VNCConnection.Settings) {
		self.initialSettings = settings
		
		super.init(nibName: "ConfigurationView", bundle: .main)
	}
	
	@available(*, unavailable)
	required init?(coder: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}
	
	override func viewDidLoad() {
		super.viewDidLoad()
		
		guard !didLoad else { return }
		didLoad = true
		
		let encodingsConfigurationView = encodingsConfigurationViewController.view
		encodingsConfigurationView.frame = placeholderViewEncodings.bounds
		
		placeholderViewEncodings.addSubview(encodingsConfigurationView)
		
		settings = initialSettings
	}
	
	override func becomeFirstResponder() -> Bool {
		textFieldHostname.becomeFirstResponder()
	}
}

extension ConfigurationViewController {
	var settings: VNCConnection.Settings {
		get {
			.init(isDebugLoggingEnabled: isDebugLoggingEnabled,
				  hostname: hostname,
				  port: port,
				  isShared: isShared,
				  isScalingEnabled: isScalingEnabled,
				  useDisplayLink: useDisplayLink,
				  inputMode: inputMode,
				  isClipboardRedirectionEnabled: isClipboardRedirectionEnabled,
				  colorDepth: colorDepth,
				  frameEncodings: frameEncodings)
		}
		set {
			isDebugLoggingEnabled = newValue.isDebugLoggingEnabled
			hostname = newValue.hostname
			port = newValue.port
			isShared = newValue.isShared
			isScalingEnabled = newValue.isScalingEnabled
			useDisplayLink = newValue.useDisplayLink
			inputMode = newValue.inputMode
			isClipboardRedirectionEnabled = newValue.isClipboardRedirectionEnabled
			colorDepth = newValue.colorDepth
			frameEncodings = newValue.frameEncodings
		}
	}
}

private extension ConfigurationViewController {
	var isDebugLoggingEnabled: Bool {
		get { checkBoxDebugLogging.state == .on }
		set { checkBoxDebugLogging.state = newValue ? .on : .off }
	}
	
	var hostname: String {
		get { textFieldHostname.stringValue }
		set { textFieldHostname.stringValue = newValue }
	}
	
	var port: UInt16 {
		get { .init(textFieldPort.integerValue) }
		set { textFieldPort.integerValue = .init(newValue) }
	}
	
	var isShared: Bool {
		get { checkBoxShared.state == .on }
		set { checkBoxShared.state = newValue ? .on : .off }
	}
	
	var isScalingEnabled: Bool {
		get { checkBoxScaling.state == .on }
		set { checkBoxScaling.state = newValue ? .on : .off }
	}
	
	var useDisplayLink: Bool {
		get { checkBoxUseDisplayLink.state == .on }
		set { checkBoxUseDisplayLink.state = newValue ? .on : .off }
	}
	
	var inputMode: VNCConnection.Settings.InputMode {
		get { .init(rawValue: .init(popupButtonInputMode.indexOfSelectedItem)) ?? .forwardKeyboardShortcutsIfNotInUseLocally }
		set { popupButtonInputMode.selectItem(at: .init(newValue.rawValue)) }
	}
	
	var isClipboardRedirectionEnabled: Bool {
		get { checkBoxClipboardRedirection.state == .on }
		set { checkBoxClipboardRedirection.state = newValue ? .on : .off }
	}
	
	var colorDepth: VNCConnection.Settings.ColorDepth {
		get {
			switch popupButtonColorDepth.indexOfSelectedItem {
				case 0:
					return .depth8Bit
				case 1:
					return .depth16Bit
				case 2:
					return .depth24Bit
				default:
					return .depth24Bit
			}
		}
		set {
			switch newValue {
				case .depth8Bit:
					popupButtonColorDepth.selectItem(at: 0)
				case .depth16Bit:
					popupButtonColorDepth.selectItem(at: 1)
				case .depth24Bit:
					popupButtonColorDepth.selectItem(at: 2)
			}
		}
	}
	
	var frameEncodings: [VNCFrameEncodingType] {
		get { encodingsConfigurationViewController.frameEncodings }
		set { encodingsConfigurationViewController.frameEncodings = newValue }
	}
}
